### PR TITLE
Add prometheus metrics for block authorship

### PR DESCRIPTION
**Pull request by emostov**
closes https://github.com/paritytech/polkadot/issues/4327

New metrics:
- `proposer_block_proposal_time`: Time to create a block for proposal. This is a superset of both `proposer_block_constructed` and `create_inherents_time`.
- `proposer_create_inherents_time`:  Time to create the inherents for the block that is about to be constructed. This is a subset of `proposer_block_proposal_time` but does _not_ overlap with `proposer_block_constructed`.

Existing metrics:
- `proposer_block_constructed`: Time to construct a block for proposal. This is a subset of `proposer_block_proposal_time` and does not overlap with `propser_create_inherents_time`
- `block_verification_time`: Time to verify a block (takes place directly prior to importing)
- `block_verification_and_import_time`: Time to verify a block and then import a block. A superset of `block_verification_time`

### CODE REVIEW

**Review by drahnr — APPROVED** [review]
LGTM besides the mentioned requrirement to use https://doc.rust-lang.org/std/time/struct.Instant.html#method.saturating_duration_since rather than `fn elapsed(..)`

File: client/basic-authorship/src/basic_authorship.rs [hunk 1]
```diff
@@ -346,10 +346,19 @@ where
...
 		let mut block_builder =
 			self.client.new_block_at(&self.parent_id, inherent_digests, PR::ENABLED)?;
 
-		for inherent in block_builder.create_inherents(inherent_data)? {
+		let create_inherents_timer = time::Instant::now();
+		let inherents = block_builder.create_inherents(inherent_data)?;
+		self.metrics.report(|metrics| {
+			metrics
+				.create_inherents_time
+				.observe(create_inherents_timer.elapsed().as_secs_f64());
```
[the remark this answers is not in the document]
Comment (emostov) [reply client/basic-authorship/src/basic_authorship.rs]: Fixed in https://github.com/paritytech/substrate/pull/10316/commits/f7ca1dfd1dc9416d8206d7537d81c8838b86877f

File: client/basic-authorship/src/basic_authorship.rs [hunk 2]
```diff
@@ -346,10 +346,22 @@ where
...
-		for inherent in block_builder.create_inherents(inherent_data)? {
+		let create_inherents_start = time::Instant::now();
+		let inherents = block_builder.create_inherents(inherent_data)?;
+		let create_inherents_end = time::Instant::now();
+
+		self.metrics.report(|metrics| {
+			metrics
+				.create_inherents_time
+				.observe(create_inherents_start
+					.saturating_duration_since(create_inherents_end).as_secs_f64());
```
Comment (bkchr) [inline client/basic-authorship/src/basic_authorship.rs:361]:
```suggestion
				.observe(create_inherents_end
					.saturating_duration_since(create_inherents_start).as_secs_f64());
```

File: client/basic-authorship/src/basic_authorship.rs [hunk 3]
```diff
@@ -529,6 +541,14 @@ where
 
 		let proof =
 			PR::into_proof(proof).map_err(|e| sp_blockchain::Error::Application(Box::new(e)))?;
+
+		let propose_with_end = time::Instant::now();
+		self.metrics.report(|metrics| {
+			metrics
+				.create_block_proposal_time
+				.observe(propose_with_start.saturating_duration_since(propose_with_end).as_secs_f64());
```
Comment (bkchr) [inline client/basic-authorship/src/basic_authorship.rs:549]:
```suggestion
				.create_block_proposal_time
				.observe(propose_with_end.saturating_duration_since(propose_with_start).as_secs_f64());
```

**Review by bkchr — APPROVED** [review]
We need to change the order, otherwise looks good.

**Comment by drahnr** [comment]
bot merge

**Pull request merged**